src/controller/middleware/auth.rs

```diff
@@ -165,3 +172,92 @@ where
         Ok(Self { user })
     }
 }
+
+// ---------------------------------------
+//
+// JWT Auth extractor from cookie
+//
+// ---------------------------------------
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct JWTCookie {
+    pub claims: auth::jwt::UserClaims,
+}
+
+#[async_trait]
+impl<S> FromRequestParts<S> for JWTCookie
+where
+    AppContext: FromRef<S>,
+    S: Send + Sync,
+{
+    type Rejection = Error;
+
+    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Error> {
+        let token = extract_token_from_cookie(parts)?;
+        let state: AppContext = AppContext::from_ref(state);
+
+        let jwt_secret = state.config.get_jwt_config()?;
+
+        match auth::jwt::JWT::new(&jwt_secret.secret).validate(&token) {
+            Ok(claims) => Ok(Self {
+                claims: claims.claims,
+            }),
+            Err(_err) => {
+                return Err(Error::Unauthorized("token is not valid".to_string()));
+            }
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct JWTCookieWithUser<T: Authenticable> {
+    pub claims: auth::jwt::UserClaims,
+    pub user: T,
+}
+
+#[async_trait]
+impl<S, T> FromRequestParts<S> for JWTCookieWithUser<T>
+where
+    AppContext: FromRef<S>,
+    S: Send + Sync,
+    T: Authenticable,
+{
+    type Rejection = Error;
+
+    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Error> {
+        let token = extract_token_from_cookie(parts)?;
+        let state: AppContext = AppContext::from_ref(state);
+
+        let jwt_secret = state.config.get_jwt_config()?;
+
+        match auth::jwt::JWT::new(&jwt_secret.secret).validate(&token) {
+            Ok(claims) => {
+                let user = T::find_by_claims_key(&state.db, &claims.claims.pid)
+                    .await
+                    .map_err(|_| Error::Unauthorized("token is not valid".to_string()))?;
+                Ok(Self {
+                    claims: claims.claims,
+                    user,
+                })
+            }
+            Err(_err) => {
+                return Err(Error::Unauthorized("token is not valid".to_string()));
+            }
+        }
+    }
+}
+
+/// Extract a token value from cookie
+///
+/// # Errors
+/// when token value from cookie is not found
+pub fn extract_token_from_cookie(parts: &Parts) -> eyre::Result<String> {
```

Comment:
are we sure we want `eyre::Result` here?

Reply:
This function returns an error if the token cannot strip the value to get only the token. 
I prefer not to return an Option<String> and return an error with a message that explains where the error came from.

you prefer Result<String, String> instead of eyre::Result?

Reply:
changed